File: app/GOExercise/GOExercise/Services/AccountLinking.swift
import Foundation

/// 友達アカウントの「バックアップ(永続化)」連携プロバイダ。
/// 匿名 uid を保持したまま Apple/Google identity を連結し、機種変・再インストールでも復旧可能にする
/// (懸念①の根治)。連携は任意で、匿名のままでも全機能使える。
enum FriendsLinkProvider: String, Sendable, CaseIterable {
    case apple
    case google

    var displayName: String {
        switch self {
        case .apple: return "Apple"
        case .google: return "Google"
        }
    }
}

/// アカウント連携の結果/エラー。
enum AccountLinkError: LocalizedError, Equatable {
    /// その Apple/Google ID が既に別アカウントに紐付いている (identity_already_exists)。
    /// マージはせず「既存アカウントに切替 / 中止」の二択を提示する。
    case alreadyLinkedToAnotherAccount
    /// サーバ側でそのプロバイダが未設定/無効 (provider_disabled 等)。誤設定の保険。
    case providerUnavailable
    /// ユーザーがキャンセルした。
    case cancelled
    /// バックエンド未接続。
    case backendUnavailable
    /// その他 (やさしい固定文で表示し、生エラーは出さない)。
    case failed

    var errorDescription: String? {
        switch self {
        case .alreadyLinkedToAnotherAccount:
            return "このアカウントは既に別のデータに紐づいています。"
        case .providerUnavailable:
            return "ただいまこの方法は利用できません。"
        case .cancelled:
            return "キャンセルされました。"
        case .backendUnavailable:
            return "友達機能につながれませんでした。"
        case .failed:
            return "バックアップに失敗しました。通信状況を確認してお試しください。"
        }
    }
}

/// 復元 (`restoreWithApple` / `restoreWithGoogle`) の成否内訳。新端末/再インストールで
/// welcome から実行する。`Store` 側でこれに `failed`/`cancelled` を足して UI へ渡す
/// ([[FriendsStore]].RestoreResult)。Apple/Google で意味は同一なので provider-neutral。
enum RestoreOutcome: Equatable, Sendable {
    /// その Apple/Google ID に紐づく既存プロフィールが見つかり、ロードした (友達/コードが戻る)。
    case restored
    /// 既存データが無く、新規アカウントとして作成した (実質サインアップ)。
    case created
}

/// Google 連携 (web/PKCE) の認可フロー。認可 URL を ASWebAuthenticationSession で開き、
/// `goexercise://` のコールバック URL を返す。View が [[GoogleSignInCoordinator]] を渡し、
/// Service 側 (Supabase) が前後で認可 URL 生成と PKCE code 交換 (`session(from:)`) を行う。
/// Apple はネイティブ id_token なので不要 (経路が違う)。
typealias WebAuthFlow = @MainActor (_ url: URL) async throws -> URL

/// 連携状態 (UI 表示用)。
struct AccountBackupStatus: Equatable, Sendable {
    /// 永続アカウントに連携済み (= 匿名でない)。
    var isBackedUp: Bool
    /// 連携済みプロバイダ名の一覧 (例: ["apple", "google"])。表示用。
    /// **複数連携時に1つだけ拾うと並び順依存で誤表示になる**(Apple なのに Google 等)ため、
    /// 全プロバイダを保持して表示側で結合する。
    var providerNames: [String]

    /// 後方互換アクセサ: 単一プロバイダ前提だった旧コード/テスト用 (先頭を返す)。
    var providerName: String? { providerNames.first }

    init(isBackedUp: Bool, providerNames: [String]) {
        self.isBackedUp = isBackedUp
        self.providerNames = providerNames
    }

    /// 後方互換イニシャライザ (単一プロバイダ指定)。
    init(isBackedUp: Bool, providerName: String?) {
        self.isBackedUp = isBackedUp
        self.providerNames = providerName.map { [$0] } ?? []
    }

    static let anonymous = AccountBackupStatus(isBackedUp: false, providerNames: [])

    /// 設定「アカウントとバックアップ」の状態文言。連携済みプロバイダを **全部** 列挙する
    /// (Apple/Google 両方連携なら「Apple・Google アカウントでバックアップ中」)。
    /// 1つだけ拾うと並び順依存で誤表示(Apple なのに Google 等)になるため。
    var backupStatusText: String {
        let names = providerNames.map(Self.providerDisplayName)
        guard !names.isEmpty else { return "アカウントでバックアップ中" }
        return "\(names.joined(separator: "・")) アカウントでバックアップ中"
    }

    static func providerDisplayName(_ provider: String) -> String {
        switch provider {
        case "apple": return "Apple"
        case "google": return "Google"
        default: return provider.capitalized
        }
    }
}

File: app-android/app/src/main/java/com/goexercise/app/data/friends/AccountLinking.kt
package com.goexercise.app.data.friends

/**
 * アカウント連携(機種変/再インストール復旧)の型。iOS `AccountLinking.swift` の移植。
 * 匿名 uid を保持したまま Apple/Google identity を連結し、復旧可能にする。連携は任意で、
 * 匿名のままでも全機能使える。**Android は iOS の鏡像**(plan §6): Google=native id_token /
 * Apple=web/PKCE(iOS は逆)。
 */
enum class FriendsLinkProvider(val displayName: String) {
    Apple("Apple"),
    Google("Google"),
}

/** アカウント連携の結果/エラー。iOS `AccountLinkError` 移植。 */
sealed class AccountLinkError(message: String) : Exception(message) {
    /** その Apple/Google ID が既に別アカウントに紐付く(identity_already_exists)。マージせず切替/中止の二択。 */
    data object AlreadyLinkedToAnotherAccount : AccountLinkError("このアカウントは既に別のデータに紐づいています。")
    /** サーバ側でそのプロバイダが未設定/無効(provider_disabled 等)。 */
    data object ProviderUnavailable : AccountLinkError("ただいまこの方法は利用できません。")
    /** ユーザーがキャンセルした。 */
    data object Cancelled : AccountLinkError("キャンセルされました。")
    /** バックエンド未接続。 */
    data object BackendUnavailable : AccountLinkError("友達機能につながれませんでした。")
    /** その他(やさしい固定文で表示し、生エラーは出さない)。 */
    data object Failed : AccountLinkError("バックアップに失敗しました。通信状況を確認してお試しください。")
}

/** 復元(restore)の成否内訳。iOS `RestoreOutcome` 移植。 */
enum class RestoreOutcome {
    /** 既存プロフィールが見つかりロードした(友達/コードが戻る)。 */
    Restored,
    /** 既存データ無し → 新規アカウントとして作成した(実質サインアップ)。 */
    Created,
}

/**
 * Apple web/PKCE の認可フロー。認可 URL を開き(Android は Custom Tabs)、`goexercise://auth-callback`
 * のコールバック URL を返す。View/coordinator が渡し、Service(Supabase)が前後で URL 生成と code 交換を行う。
 * iOS `WebAuthFlow`(typealias)の移植。Google は native id_token なので不要(経路が違う)。
 */
typealias WebAuthFlow = suspend (authUrl: String) -> String

/** 連携状態(UI 表示用)。iOS `AccountBackupStatus` 移植。 */
data class AccountBackupStatus(
    /** 永続アカウントに連携済み(= 匿名でない)。 */
    val isBackedUp: Boolean,
    /**
     * 連携済みプロバイダ名の一覧(例: ["apple", "google"])。表示用。
     * **複数連携時に1つだけ拾うと並び順依存で誤表示になる**(Apple なのに Google 等)ため、
     * 全プロバイダを保持して表示側で結合する(iOS 1.3 パリティ)。
     */
    val providerNames: List<String> = emptyList(),
) {
    /** 後方互換アクセサ: 単一プロバイダ前提だった旧コード/テスト用(先頭を返す)。 */
    val providerName: String? get() = providerNames.firstOrNull()

    /** 単一プロバイダ指定の後方互換コンストラクタ。 */
    constructor(isBackedUp: Boolean, providerName: String?) :
        this(isBackedUp, providerName?.let { listOf(it) } ?: emptyList())

    /**
     * 設定「アカウントとバックアップ」の状態文言。連携済みプロバイダを **全部** 列挙する
     * (Apple/Google 両方連携なら「Apple・Google アカウントでバックアップ中」)。
     * 1つだけ拾うと並び順依存で誤表示(Apple なのに Google 等)になるため。
     */
    val backupStatusText: String
        get() {
            val names = providerNames.map { providerDisplayName(it) }
            return if (names.isEmpty()) "アカウントでバックアップ中"
            else "${names.joinToString("・")} アカウントでバックアップ中"
        }

    /** 連携済みプロバイダの表示名を「・」で結合(未連携は null)。設定の「✓ X で連携済み」用。 */
    val linkedProvidersDisplay: String?
        get() = providerNames.takeIf { it.isNotEmpty() }?.joinToString("・") { providerDisplayName(it) }

    companion object {
        val Anonymous = AccountBackupStatus(isBackedUp = false, providerNames = emptyList())

        fun providerDisplayName(provider: String): String = when (provider) {
            "apple" -> "Apple"
            "google" -> "Google"
            else -> provider.replaceFirstChar { it.uppercase() }
        }
    }
}
